include/sacn/cpp/dmx_merger.h

```diff
@@ -0,0 +1,336 @@
+{
+  source_id_t result = SACN_DMX_MERGER_SOURCE_INVALID;
+  etcpal_error_t err = sacn_dmx_merger_add_source(handle_, &source_cid.get(), &result);
+  if (err == kEtcPalErrOk)
+    return result;
+  else
+    return err;
+}
+
+/*!
+ * \brief Removes a source from the merger.
+ *
+ * Removes the source from the merger.  This causes the merger to recalculate the outputs.
+ *
+ * \param[in] source The id of the source to remove.
+ * \return #kEtcPalErrOk: Source removed successfully.
+ * \return #kEtcPalErrInvalid: Invalid parameter provided.
+ * \return #kEtcPalErrNotInit: Module not initialized.
+ * \return #kEtcPalErrSys: An internal library or system call error occurred.
+ */
+inline etcpal::Error DmxMerger::RemoveSource(source_id_t source)
+{
+  return sacn_dmx_merger_remove_source(handle_, source);
+}
+
+/*!
+ * \brief Returns the source id for that source cid.
+ *
+ * \param[in] source_cid The UUID of the source CID.
+ * \return On success this will be the source ID, otherwise kEtcPalErrInvalid.
+ */
+inline etcpal::Expected<source_id_t> DmxMerger::GetSourceId(const etcpal::Uuid& source_cid) const
+{
+  source_id_t result = sacn_dmx_merger_get_id(handle_, &source_cid.get());
+  if (result != SACN_DMX_MERGER_SOURCE_INVALID)
+    return result;
+  return kEtcPalErrInvalid;
+}
+
+/*!
+ * \brief Gets a read-only view of the source data.
+ *
+ * Looks up the source data and returns a pointer to the data or nullptr if it doesn't exist.
+ * This pointer is owned by the library, and must not be modified by the application.
+ * The pointer will only be valid until the source or merger is removed.
+ *
+ * \param[in] source The id of the source.
+ * \return The reference to the source data, otherwise kEtcPalErrInvalid.
+ */
+inline const SacnDmxMergerSource* DmxMerger::GetSourceInfo(source_id_t source) const
+{
+  return sacn_dmx_merger_get_source(handle_, source);
+}
+
+/*!
+ * \brief Updates the source data and recalculate outputs.
+ *
+ * The direct method to change source data.  This causes the merger to recalculate the outputs.
+ * If you are processing sACN packets, you may prefer UpdateSourceDataFromSacn().
+ *
+ * \param[in] source The id of the source to modify.
+ * \param[in] new_values The new DMX values to be copied in. This may be nullptr if the source is only updating the
+ * priority or address_priorities.
+ * \param[in] new_values_count The length of new_values. May be 0 if the source is only updating the priority or
+ * address_priorities.
+ * \param[in] priority The universe-level priority of the source.
+ * \param[in] address_priorities The per-address priority values to be copied in.  This may be nullptr if the source is
+ * not sending per-address priorities, or is only updating other parameters.
+ * \param[in] address_priorities_count The length of address_priorities.  May be 0 if the source is not sending these
+ * priorities, or is only updating other parameters.
+ * \return #kEtcPalErrOk: Source updated and merge completed.
```

Comment:
My original idea for the api was to allow both models.  The update_source functions originally took a dont_calculate_now flag, and then recalculate would cause the calculation.  But the reference code did it every time and it hadn't been a problem.  It would be more computationally expensive to do this for all sources than for a single source.  @ChristianReese what do you think?

Reply:
If we separated this out, the merge function in the recalculate function would not know about the specific source being updated, and would lose some optimizations that utilized that (it would go through each source for each slot every time). Maybe there's a design to track the changed sources and optimize that way? However, the performance savings may outweigh the losses depending on the use case.

Implementing this would not at all be difficult. It could be a boolean "calculate_later" flag or just a separate function. Everything would pretty much be the same in the code except it wouldn't call the merge function. Creating a source-less merge function for recalculate should also be pretty quick.

Reply:
I've been thinking about the idea of tracking the sources that changed. Then we could just feed them into the same merge function we're using now. Then the optimizations could be preserved, and I can't think of gotchas that would actually make it worse. 

The optimization, specifically, is that it only does a full recalculation (iterating all sources) for a slot if the source being merged in was the last known winner, and it may have lost precedence. Otherwise, if it beats the last known winner, the winner values are updated without iterating through other sources. Otherwise, nothing happens.

The design would be that the update functions would not call the merge function, but would keep another rb-tree of the sources that changed (handle/pointer pairs). Then recalculate would iterate that rb-tree and call the existing merge function for each one. No new merge function needed.

Reply:
How much memory are we adding?

Reply:
And I think, in this design, a call to the recalculate function would only ever hit the most expensive "may have lost precedence" case once. So it would be: for each slot, for each changed source, call merge_source, which will only be O(n) for one of those changed sources, and O(1) for the others. This would only be significantly different from a full recalculate if the changed sources were a small percentage of all the sources. Otherwise it may end up being a bit slower (but only by a constant factor).

Reply:
@nwagner Didn't see your comment for some reason. The memory would only be one red-black tree per merger containing 48-byte entries for each changed source. so 48 bytes * MAX_SOURCES_PER_MERGER * MAX_MERGERS.

Reply:
This would give us more flexibility.  Would this be less complex if they can only set up the mode when they create the instance of the merger, rather than having a calculate_later flag on each call?

Reply:
I still see the merge_receiver merging on each sACN packet update, however.

Reply:
I guess I'm ok with it if you want to make those changes..  You can make the changes to the c++ header after it's merged as well. Sam, what do you think?

Reply:
I don't feel informed enough on this implementation to offer much of an opinion. It may be that it's a premature optimization that we don't have to make. I was just a bit worried about this being called directory from the packet reception context, and thus block the receipt of more packets. If it's not an issue in the reference code, maybe it's fine.

If we _do_ do it, I would recommend against just a boolean flag for recalculation (it can cause the code to be more obfuscated at the point of use). Instead I would either use an enum:
```cpp
enum class RecalculateBehavior
{
  RecalculateNow,
  DeferRecalculation
};
```

Or to have two functions, `UpdateSourceData()` and `UpdateSourceDataAndRecalculate()`.

Reply:
Ultimately, I pulled it from the API because it seemed fine in the reference code.  We could leave it as-is and see if anyone remarks upon it at the future API review. 
I like the idea of an enum if we do go this route.

Reply:
I'm good with that. Keep it simple now, add complexity later if needed.